### PR TITLE
Add Custom Provider for SecureAuth IdP

### PR DESCRIPTION
**Pull request by markafarrell**
# Overview
SecureAuth IdP returns group membership as a comma separated  list of strings (e.g. `groups: "group-1,group-2"`) rather than a JSON list.
This custom provider converts this into a JSON list so that the entity is added to the correct groups on login.

# Design of Change
The custom `GroupFetcher` simply splits the `GroupClaims` field from the token on comma and returns the resultant list of groups 

# Related Issues/Pull Requests
N/A

# Contributor Checklist
[ ] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
[My Docs PR Link](link)
[Example](https://github.com/hashicorp/vault/commit/2715f5cec982aabc7b7a6ae878c547f6f475bba6)
[x] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
[x] Backwards compatible

### CODE REVIEW

**Comment by markafarrell** [comment]
@austingebauer Is something additional required to have this reviewed? It seems there are a few PRs with no reviews or comments since early 2021.

**Comment by austingebauer** [comment]
Hi, @markafarrell! Thanks for this contribution and for bringing it to my attention. I'm going to discuss the possibility of getting this merged with my team. The changes in this PR generally look good (interesting that they return the groups in a single string 😄). We'd like to be able to test the plugin using SecureAuth IdP before documenting it along with other tested [OIDC providers](https://www.vaultproject.io/docs/auth/jwt/oidc_providers). 

I'll get back to you with an update after having the discussion. Thanks again!

**Comment by markafarrell** [comment]
Thanks @austingebauer for the prompt response.

I think there are a few other stale PRs that deserve some love. Especially #130

**Comment by austingebauer** [comment]
Hi @markafarrell - Wanted to give you an update that we were able to get a SecureAuth IdP to test again. Things are looking good! I'm going to submit a review for this shortly and we'll get it merged. Thanks again for this contribution and the patience.

File: provider_secureauth.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			}
+			groupsClaimRaw = secureauthGroups
+		}
+	}
+	b.Logger().Debug(fmt.Sprintf("post: groups claim raw is %v", groupsClaimRaw))
+	return groupsClaimRaw, nil
+}
+
+
+type secureauthGroups struct {
```
Comment (austingebauer) [inline provider_secureauth.go:51]: Looks unused?

File: provider_secureauth.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// SensitiveKeys - satisfying the CustomProvider interface
+func (a *SecureAuthProvider) SensitiveKeys() []string {
+	return []string{}
+}
+
+// FetchGroups - custom groups fetching for secureauth - satisfying GroupsFetcher interface
+// SecureAuth by default will return groups not as a json list but as a list of comma seperated strings
+// We need to convert this to a json list
+
+func (a *SecureAuthProvider) FetchGroups(_ context.Context, b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole, tokenSource oauth2.TokenSource) (interface{}, error) {
```
Comment (austingebauer) [inline provider_secureauth.go:31]: I think you can use `_` for `tokenSource` as well.

File: provider_secureauth.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+// SensitiveKeys - satisfying the CustomProvider interface
+func (a *SecureAuthProvider) SensitiveKeys() []string {
+	return []string{}
+}
+
+// FetchGroups - custom groups fetching for secureauth - satisfying GroupsFetcher interface
+// SecureAuth by default will return groups not as a json list but as a list of comma seperated strings
+// We need to convert this to a json list
+
```
Comment (austingebauer) [inline provider_secureauth.go:30]:
```suggestion
```

File: provider_secureauth_test.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
+package jwtauth
```
Comment (austingebauer) [inline provider_secureauth_test.go:1]: Thanks for writing the tests!

File: provider_secureauth.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	"fmt"
+	"strings"
+
+	"golang.org/x/oauth2"
+)
+
+// SecureAuthProvider is used for SecureAuth-specific configuration
+type SecureAuthProvider struct {
+	// Context for secureauth calls
+	ctx context.Context
```
Comment (austingebauer) [inline provider_secureauth.go:14]: This looks unused. Can likely remove it.

**Review by austingebauer — COMMENTED** [review]
@markafarrell - I tested this out, and everything works as expected. I have a few minor suggestions. We'll merge this once those are addressed.

**Comment by markafarrell** [comment]
@austingebauer I think I have addressed all your comments. I'm on parental leave from work at the moment so I'm not able to test the changes end-to-end though. Hopefully your setup allows you to do this.

P.S. I'm getting an issue with `make bin` using `go version go1.18.3 linux/amd64`

```
1 errors occurred:
--> darwin/386 error: exit status 2
Stderr: go: unsupported GOOS/GOARCH pair darwin/386
```

It appears that >= go 1.15  doens't support darwin/386 anymore (https://go.dev/doc/go1.15),  however, trying to build the plugin with `go version go1.14.15 linux/amd64` I get issues with `golang.org/x/net/http2` `note: module requires Go 1.17` and `github.com/hashicorp/vault/sdk/helper/certutil` `note: module requires Go 1.16`

Disabling the `darwin/386` target in `scripts/build.sh` fixes the issue though.

**Comment by markafarrell** [comment]
Thanks for all the help getting this merged @austingebauer !

**Pull request merged**